File: 1747-lexicographically-smallest-string-after-applying-operations/lexicographically-smallest-string-after-applying-operations.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string addOperation(string s, int a) {
        for (int i = 1; i < s.size(); i += 2) {
            s[i] = ((s[i] - '0' + a) % 10) + '0';
        }
        return s;
    }

    string rotateOperation(string s, int b) {
        int n = s.size();
        return s.substr(n - b) + s.substr(0, n - b);
    }

    string findLexSmallestString(string s, int a, int b) {
        queue<string> q;
        unordered_set<string> seen;
        string ans = s;

        q.push(s);
        seen.insert(s);

        while (!q.empty()) {
            string cur = q.front();
            q.pop();

            if (cur < ans)
                ans = cur;

            // Operation 1: add
            string addStr = addOperation(cur, a);
            if (!seen.count(addStr)) {
                seen.insert(addStr);
                q.push(addStr);
            }

            // Operation 2: rotate
            string rotStr = rotateOperation(cur, b);
            if (!seen.count(rotStr)) {
                seen.insert(rotStr);
                q.push(rotStr);
            }
        }
        return ans;
    }
};

File: depth-first-search/lexicographically-smallest-string-after-applying-operations.java
import java.util.*;

class Solution {
    public String findLexSmallestString(String s, int a, int b) {
        Queue<String> queue = new LinkedList<>();
        Set<String> seen = new HashSet<>();
        String smallest = s;

        queue.add(s);
        seen.add(s);

        while (!queue.isEmpty()) {
            String cur = queue.poll();
            if (cur.compareTo(smallest) < 0) {
                smallest = cur;
            }

            // Operation 1: add 'a' to all odd indices
            String added = addOperation(cur, a);
            if (seen.add(added)) {
                queue.add(added);
            }

            // Operation 2: rotate by 'b'
            String rotated = rotateOperation(cur, b);
            if (seen.add(rotated)) {
                queue.add(rotated);
            }
        }

        return smallest;
    }

    // Add 'a' to all digits at odd indices
    private String addOperation(String s, int a) {
        char[] chars = s.toCharArray();
        for (int i = 1; i < chars.length; i += 2) {
            int num = (chars[i] - '0' + a) % 10;
            chars[i] = (char) (num + '0');
        }
        return new String(chars);
    }

    // Rotate string to the right by 'b' positions
    private String rotateOperation(String s, int b) {
        int n = s.length();
        return s.substring(n - b) + s.substring(0, n - b);
    }
}
